pkg/services/store/object/http.go

```diff
@@ -0,0 +1,191 @@
+	route.Get("/history/*", reqGrafanaAdmin, routing.Wrap(s.doGetHistory))
+	route.Get("/list/*", reqGrafanaAdmin, routing.Wrap(s.doListFolder)) // Simplified version of search -- path is prefix
+	route.Get("/search", reqGrafanaAdmin, routing.Wrap(s.doSearch))
+}
+
+// This function will extract UID+Kind from the requested path "*" in our router
+// This is far from ideal! but is at least consistent for these endpoints.
+// This will quickly be revisited as we explore how to encode UID+Kind in a "GRN" format
+func parseRequestParams(req *http.Request) (uid string, kind string, params map[string]string) {
+	params = web.Params(req)
+	path := params["*"]
+	idx := strings.LastIndex(path, ".")
+	if idx > 0 {
+		uid = path[:idx]
+		kind = path[idx:]
+	} else {
+		uid = path
+		kind = "?"
+	}
+	return
+}
+
+func (s *httpObjectStore) doGetObject(c *models.ReqContext) response.Response {
+	uid, kind, params := parseRequestParams(c.Req)
+	rsp, err := s.store.Read(c.Req.Context(), &ReadObjectRequest{
+		UID:         uid,
+		Kind:        kind,
+		Version:     params["version"], // ?version = XYZ
+		WithBody:    true,              // ?? allow false?
+		WithSummary: true,              // ?? allow false?
+	})
+	if err != nil {
+		return response.Error(500, "error fetching object", err)
+	}
+	if rsp.Object == nil {
+		return response.Error(404, "not found", nil)
+	}
+
+	// Configure etag support
+	currentEtag := rsp.Object.ETag
+	previousEtag := c.Req.Header.Get("If-None-Match")
+	if previousEtag == currentEtag {
+		return response.CreateNormalResponse(
+			http.Header{
+				"ETag": []string{rsp.Object.ETag},
+			},
+			[]byte{},               // nothing
+			http.StatusNotModified, // 304
+		)
+	}
+
+	c.Resp.Header().Set("ETag", currentEtag)
+	return response.JSON(200, rsp)
+}
+
+func (s *httpObjectStore) doGetRawObject(c *models.ReqContext) response.Response {
+	uid, kind, params := parseRequestParams(c.Req)
+	rsp, err := s.store.Read(c.Req.Context(), &ReadObjectRequest{
+		UID:         uid,
+		Kind:        kind,
+		Version:     params["version"], // ?version = XYZ
+		WithBody:    true,
+		WithSummary: false,
+	})
+	if err != nil {
+		return response.Error(500, "?", err)
+	}
+	if rsp.Object != nil && rsp.Object.Body != nil {
+		// Configure etag support
+		currentEtag := rsp.Object.ETag
+		previousEtag := c.Req.Header.Get("If-None-Match")
+		if previousEtag == currentEtag {
+			return response.CreateNormalResponse(
+				http.Header{
+					"ETag": []string{rsp.Object.ETag},
+				},
+				[]byte{},               // nothing
+				http.StatusNotModified, // 304
+			)
+		}
+
+		return response.CreateNormalResponse(
+			http.Header{
+				"Content-Type": []string{"application/json"}, // TODO, based on kind!!!
+				"ETag":         []string{currentEtag},
+			},
+			rsp.Object.Body,
+			200,
+		)
+	}
+	return response.JSON(400, rsp) // ???
+}
+
+func (s *httpObjectStore) doWriteObject(c *models.ReqContext) response.Response {
```

Comment:
this looks like the place for rate limiting/limiting the file size, do we want to add that in this PR?

Reply:
🤔 I guess good to add at the HTTP level so it can avoid a large upload -- however, if we have real rate/size limits, they should be in the store itself (i think)

Also not sure what reasonable limits are... dashboards/images can be big -- maybe we set to 5MB for now?

Reply:
file size check on the HTTP level is nice cause it means we don't need to read the whole file (e.g. 100mb) via ` := io.ReadAll(c.Req.Body)` before rejecting it. Ultimately its probably needed in both places - the maximum allowable size in HTTP, and configurable/reasonable limit in the service

Reply:
I'll hard code a 5MB limit for now -- and we can revisit